book/src/proposals/booting-from-snapshot.md

```diff
@@ -0,0 +1,316 @@
+If it gets a snaphsot for slot 200, and it can see that all of its votes are in
+its vote account for slot 200, it knows that 200 must be a descendent of 100,
+so it can safely vote on 200 and its descendents. However, assume that the
+cluster is at 300 when the validator gets the snapshot for 200, and the
+validator replays and votes on slots 200-250 while catching up, but then
+reboots again before catching up. When the validator reboots again, assume it
+gets a snapshot for slot 300. It will not see its votes for slots 200-250 in
+its vote account in bank 300, so it will have to assume that they were for
+another fork, and wait for their lockouts to expire accordingly. Avoiding this
+situation requires the caught up check and the correct cluster check (a wrong
+cluster could also be missing the votes).
+
+
+## Snapshot Verification Overview
+
+A booting validator needs to have some set of trusted validators whose votes it
+will rely upon to ensure that it has caught up to the cluster and has a valid
+snapshot. With that set, the validator can follow the procedure detailed in the
+`Boot Procedure` section.
+
+For the sections below:
+
+Assume an arbitrary `Tower` state `tower`, a `snapshot_root`, and a
+`trusted_bank` (defined in step 8-1 of the `Boot Procedure` section).
+
+Define `tower_vote_slot` to be: The slot of some vote in `tower`.
+
+Define the function `is_ancestor(a, b)` to be: Given a slot `a` and a slot `b`
+will return true if `a` is an ancestor of `b`.
+
+Define the function `is_locked(a, b)` to be: `a + a.lockout <= b`
+
+### Additions to a Snapshot
+
+The snapshot is augmented to store the last `N` ancestors of `snapshot_root`.
+These ancestors are incorporated into the bank hash so they can be verified
+when a validator unpacks a snapshot. Call this set `ancestors`.
+
+A set of votes that can prove lockout on `snapshot_root` is also added to the
+snapshot package. Call this set `snapshot_votes`. The validator generating the
+package will include a set of votes from each validator in the leader schedule.
+For each included validator, the set of votes will be the votes on the fork
+that achieves the highest lockout on `snapshot_root` (up to
+`MAX_LOCKOUT_HISTORY`). The complete vote transaction will be included so that
+the signature is verifiable.  If available, a vote for `snapshot_root` must be
+included to verify `snapshot_root`'s bank hash.
+
+
+## Boot Procedure:
+
+1) The validator gets a snapshot to boot from. This can come from any source,
+as it will be verified before being vote upon.
+
+2) The validator rejects any snapshot where `snapshot_root` <= the largest root
+recorded in blocktree.
+
+3) The validator rejects any snapshot where `tower.root` is not an ancestor of
+`snapshoot_root`.  If `tower.root` < `snapshot_root`, this can be done by
+checking if `tower.root` is contained in the snapshot's `ancestors` set.  If
+`tower.root` == `snapshot_root` then ancestry is proven. `tower.root` <=
```

Comment:
I think we should say "It is guaranteed `tower.root` <= ... if these conditions hold: " . Otherwise this statement seems out of place.

Reply:
Also, I think it would be clearer to move the `tower.root <=` part into part 2) and change part 2) to read 

a) "It is guaranteed tower.root <= snapshot_root ... iff these conditions hold"
c) We reject any snapshot_root < tower.root
b) Thus, we know `snapshot_root` <= `tower.root`